### PR TITLE
Support for PXE server on HIMN

### PR DESCRIPTION
**Pull request by simonjbeaumont**

### CODE REVIEW

File: ocaml/xapi/xapi_network.ml [hunk 1]
```diff
@@ -29,14 +29,16 @@ let create_internal_bridge ~__context ~bridge ~uuid =
...
-	if not(List.mem_assoc "ip_begin" other_config) then
-		error "Cannot setup host internal management network: no other-config:ip_begin"
-	else begin
-		(* Set the ip address of the bridge *)
+	let open Network_interface in
+	let dbg = Context.string_of_task __context in
+	try
 		let ip = List.assoc "ip_begin" other_config in
-		ignore(Forkhelpers.execute_command_get_output "/sbin/ifconfig" [bridge; ip; "up"]);
-		Xapi_mgmt_iface.enable_himn ~__context ~addr:ip
```
Comment (robhoes) [inline ocaml/xapi/xapi_network.ml]: Did you mean to remove this line?

File: ocaml/xapi/xapi_network.ml [hunk 2]
```diff
@@ -20,23 +20,26 @@ open D
...
-let create_internal_bridge ~__context ~bridge ~uuid =
+let create_internal_bridge ~__context ~bridge ~uuid ~persist =
 	debug "Creating internal bridge %s (uuid:%s)" bridge uuid;
 	let dbg = Context.string_of_task __context in
 	let current = Net.Bridge.get_all dbg () in
 	if not(List.mem bridge current) then
 		let other_config = ["network-uuids", uuid] in
-		Net.Bridge.create dbg ~name:bridge ~other_config ()
+		Net.Bridge.create dbg ~name:bridge ~other_config ();
+	Net.Bridge.set_persistent dbg ~name:bridge ~value:persist
```
Comment (robhoes) [inline ocaml/xapi/xapi_network.ml]: The indentation suggests that you intended this statement to be outside the `then` clause of the `if` statement above (i.e. always execute it). This is not the case, however: the `let` binding on line 28 sucks all the semicolon-separated statements below it into the `then` clause.

In a top-level, try this:
```
if false then let () = () in print_endline "A"; print_endline "B";;
```
vs.
```
if false then print_endline "A"; print_endline "B";;
```

Or perhaps you knew this and the indentation is wrong?

File: ocaml/xapi/xapi_network.ml [hunk 3]
```diff
@@ -20,23 +20,26 @@ open D
...
 	let open Network_interface in
 	let dbg = Context.string_of_task __context in
 	try
 		let ip = List.assoc "ip_begin" other_config in
 		let netmask = List.assoc "netmask" other_config in
+		let persist = try List.assoc "persist" other_config |> bool_of_string with _ -> false in
 		let ipv4_conf =
 			(Static4 [Unix.inet_addr_of_string ip, netmask_to_prefixlen netmask]) in
 		Net.Interface.set_ipv4_conf dbg bridge ipv4_conf;
+		if persist then Net.Interface.set_persistent dbg bridge true;
```
Comment (robhoes) [inline ocaml/xapi/xapi_network.ml]: Perhaps just

    Net.Interface.set_persistent dbg bridge persist;

This way, you don't rely on networkd's default for the `persistent` field.

File: ocaml/xapi/xapi_udhcpd.ml [hunk 4]
```diff
@@ -154,75 +157,71 @@ let restart_nolock () =
...
-	Helpers.log_exn_continue (Printf.sprintf "maybe_add_lease VIF:%s" (Ref.string_of vif)) 
-		(fun () ->
-			Mutex.execute mutex
-				(fun () ->
-					maybe_add_lease_nolock ~__context vif
-				)
-		) ()
+	Helpers.log_exn_continue (Printf.sprintf "maybe_add_lease VIF:%s" (Ref.string_of vif)) (fun () ->
+		Mutex.execute mutex (fun () ->
+				maybe_add_lease_nolock ~__context vif
```
Comment (robhoes) [inline ocaml/xapi/xapi_udhcpd.ml]: Lots of tabs here.

File: ocaml/xapi/xapi_network.ml [hunk 5]
```diff
@@ -20,23 +20,28 @@ open D
...
 		let ip = List.assoc "ip_begin" other_config in
-		ignore(Forkhelpers.execute_command_get_output "/sbin/ifconfig" [bridge; ip; "up"]);
-		Xapi_mgmt_iface.enable_himn ~__context ~addr:ip
-	end
+		let netmask = List.assoc "netmask" other_config in
+		let persist = try List.assoc "persist" other_config |> bool_of_string with _ -> false in
+		let ipv4_conf =
+			(Static4 [Unix.inet_addr_of_string ip, netmask_to_prefixlen netmask]) in
+		Net.Interface.set_ipv4_conf dbg bridge ipv4_conf;
+		if persist then Net.Interface.set_persistent dbg bridge true;
```
Comment (robhoes) [inline ocaml/xapi/xapi_network.ml]: Perhaps just

    Net.Interface.set_persistent dbg bridge persist;

This way, you don't rely on networkd's default for the `persistent` field.

File: ocaml/xapi/xapi_network.ml (hunk 2 above)
Comment (simonjbeaumont) [inline ocaml/xapi/xapi_network.ml]: Good spot! 

> Or perhaps you knew this and the indentation is wrong?
Nope. I thought the if bound to a `;`... that was an oversight that `let` would gobble up forever.

File: ocaml/xapi/xapi_network.ml (hunk 1 above)
Comment (simonjbeaumont) [inline ocaml/xapi/xapi_network.ml]: Probably not ;)

File: ocaml/xapi/xapi_network.ml (hunk 5 above)
Comment (simonjbeaumont) [inline ocaml/xapi/xapi_network.ml]: That's true, would you consider that preferable to avoiding an RPC to networkd?

File: ocaml/xapi/xapi_udhcpd.ml (hunk 4 above)
Comment (simonjbeaumont) [inline ocaml/xapi/xapi_udhcpd.ml]: I'll fix that up.

File: ocaml/xapi/xapi_network.ml (hunk 5 above)
Comment (robhoes) [inline ocaml/xapi/xapi_network.ml]: I would, yes.

File: ocaml/xapi/xapi_network.ml (hunk 1 above)
Comment (simonjbeaumont) [inline ocaml/xapi/xapi_network.ml]: Fixed by a9532e4

File: ocaml/xapi/xapi_network.ml (hunk 2 above)
Comment (simonjbeaumont) [inline ocaml/xapi/xapi_network.ml]: Fixed by 674f675

File: ocaml/xapi/xapi_udhcpd.ml (hunk 4 above)
Comment (simonjbeaumont) [inline ocaml/xapi/xapi_udhcpd.ml]: Fixed by fba2108

File: ocaml/xapi/xapi_network.ml (hunk 5 above)
Comment (simonjbeaumont) [inline ocaml/xapi/xapi_network.ml]: Fixed by 4dcfb2d

**Comment by simonjbeaumont** [comment]
@robhoes I've added fixup commits for all of your comments. when you're happy let me know and I'll squash them in.

**Comment by robhoes** [comment]
:shipit:

**Comment by simonjbeaumont** [comment]
I've squashed the `fixup` commits above before merging.

**Pull request merged**